File: src/main/java/org/algo4j/linear/IntPriorityQueue.java
package org.algo4j.linear;

import org.algo4j.tree.heap.MinHeap;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;

/**
 * Created by ice1000 on 2017/1/27.
 *
 * @author ice1000
 */
@SuppressWarnings("WeakerAccess")
public class IntPriorityQueue extends MinHeap implements IntContainer {
	protected IntPriorityQueue(@NotNull int... data) {
		super(data);
	}

	@NotNull
	@Contract("_ -> !null")
	public static IntPriorityQueue make(@NotNull int... origin) {
		return new IntPriorityQueue(MinHeap.makeHeap(origin));
	}

	@NotNull
	@Override
	public IntPriorityQueue clone() {
		return new IntPriorityQueue(data);
	}
}

File: src/main/kotlin/org/algo4j/IntPriorityQueue.kt
@file:JvmMultifileClass
@file:JvmName("Extension")

package org.algo4j

import org.algo4j.linear.IntPriorityQueue
import org.jetbrains.annotations.Contract

/**
 * Created by ice1000 on 2017/1/27.
 *
 * @author ice1000
 */


@Contract(pure = true)
operator fun IntPriorityQueue.get(index: Int) = getElementAt(index)

@Contract(pure = false)
operator fun IntPriorityQueue.plusAssign(value: Int) =
		push(value)

/**
 * functional plus
 */
@Contract(pure = true)
operator fun IntPriorityQueue.plus(value: Int) = clone().apply { plusAssign(value) }
